rkt: add experimental support for attachable applications

lucab opened the pull request:
This commit introduces experimental support for attachable
applications. It consists of:
 * a new `attach` subcommand
 * a set of per-app flags to control stdin/stdout/stderr modes
 * a stage1 `iottymux` binary for multiplexing and attaching
 * two new templated stage1 services, `iomux` and `ttymux`

Fixes https://github.com/coreos/rkt/issues/1799

---

TODO:
 * [x] rebase on master once https://github.com/coreos/rkt/pull/3388 is merged
 * [x] cleanup iottymux code, many TODO pending in there
 * [x] add developer documentation, as the whole mechanism is complex
 * [x] add user documentation for `rkt attach` and `--stdin`, `--stdout`, `--stderr`
 * [x] add smoke tests for stream and tty-attach
 * [ ] ungate, this is currently behind a `$RKT_EXPERIMENT_ATTACH` flag
 * [ ] signal proxying and custom detach

squeed commented:
Hm.
Does this mean you need to decide at `app add` whether or not you are going to attach a TTY? Will this make kubernetes mad?

squeed commented:
Oh, nevermind, I get it! `tty` is a remote interactive process, whereas `interactive` is a local interactive process.

squeed commented on stage1/init/common/units.go:
```diff
@@ -179,6 +171,131 @@ func ImmutableEnv(p *stage1commontypes.Pod, interactive bool, privateUsers strin
 
+// SetupAppIO prepares all properties related to streams (stdin/stdout/stderr) and TTY
+// for an application service unit
+func (w *UnitWriter) SetupAppIO(p *stage1commontypes.Pod, ra *schema.RuntimeApp, binPath string, interactive bool, debug bool, opts ...*unit.UnitOption) []*unit.UnitOption {
```
I know this is a WIP but I would love this to have a really good docblock, plus a reference to whatever other docs you write. Being nice to the future programmer who comes and wonders what the heck we're doing.

squeed commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+
+	ec := make(chan error)
+	for s, p := range eps {
+		conn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%s", p))
```
Why did you choose to use a network socket instead of a Unix socket?

lucab replied:
No strong design decision, but network sockets look more flexible here as they are already scoped per-pod, they are not tied to the overabundant mount-ns magic we have in place and can play nicely with nsenter (if needed).

lucab replied:
Slowly starting to re-consider this choice in favor of a stage1-namespaced file-backed unix socket :blush:

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+	}
+	// Open FIFOs
+	var endpoints = make([]*net.Listener, 3)
+	var fifos = make([]*os.File, 3)
```
why do we need a slice here?

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+func actionIOMux(statusFile *os.File) error {
+	var err error
+	var endpoints = make([]*net.Listener, 3)
+	var fifos = make([]*os.File, 3)
```
let's document the nature of these slices -> `stdin, stdout, stderr`

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+}
+
+// proxyIO performs bi- or mono-directional byte-by-byte forwarding
+func proxyIO(clients chan *net.Conn, tty *os.File, hasIn bool, hasOut bool) {
```
`hasIn, hasOut bool`

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+	var logs []io.WriteCloser
+	var alivel []bool
+	var conns []io.WriteCloser
+	var alivec []bool
```
instead of maintaining `alivec` and `alivel` slices, you could remove them simply from the `logs` and `conns` slices in cases errors occur.

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,535 @@
+}
+
+// acceptConn accepts a single client and queues it for further proxying
+func acceptConn(socket *net.Listener, c chan *net.Conn, stream string) {
```
`net.Listener` is an interface, please don't pass a pointer to it.

s-urbaniak replied:
`net.Conn` is an interface, please don't send pointers of it via a channel, it boils down to being a two-word pair of a pointer and type already, see http://research.swtch.com/interfaces.

lucab commented on rkt/rkt.go:
```diff
@@ -389,3 +390,25 @@ func getConfig() (*config.Config, error) {
+
+// getRktEnvExperiments returns annotations for rkt experimental/hidden
+// features
+func getRktEnvExperiments(annotations map[types.ACIdentifier]string) map[types.ACIdentifier]string {
```
I removed this code, will reconsider these features later in a better way.

lucab commented on rkt/attach.go:
```diff
@@ -0,0 +1,181 @@
+	defer p.Close()
+
+	if p.State() != pkgPod.Running {
+		stderr.Printf("pod %q isn't currently running", p.UUID)
```
Everywhere else we bubble up the return code to the wrapper and I prefer to stick to it. It may make sense to return the error object instead of the exit code, but not in the scope of this PR.

euank replied:
👍 for bubbling up so defers happen; I don't think it needs or should be an error object either necessarily since one step up won't have more info in this case ever.

euank commented on rkt/attach.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	if err = stage0.Attach(p.Path(), podPID, *appName, stage1RootFS, uuid, attachArgs); err != nil {
+		stderr.PrintE("enter failed", err)
```
'attach' might be clearer here

lucab replied:
ack

euank commented on rkt/attach.go:
```diff
@@ -0,0 +1,181 @@
+	attachArgs := []string{
+		fmt.Sprintf("--app=%s", appName),
+		fmt.Sprintf("--debug=%t", debug),
+	}
```
Renaming to "createAttachFlags" might help

lucab replied:
You are right, this func is misleading. I'll split the concerns, move name and debug out of here, and rename it.

euank commented on rkt/attach.go:
```diff
@@ -0,0 +1,181 @@
+		return nil, fmt.Errorf("incompatible endpoints %q, cannot simultaneously attach TTY and streams", attachMode)
+	}
+
+	attachArgs = append(attachArgs, []string{
```
we can drop the `[]string{}...` bit, right?

lucab replied:
Definitely, silly me :(

euank commented on stage1/attach/attach.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	args := stage1common.PrepareEnterCmd(false)
+	args = append(args, []string{
```
The `[]string{}...` is redundant

lucab commented on rkt/attach.go:
```diff
@@ -0,0 +1,181 @@
+		attachTTYIn = true
+		attachTTYOut = true
+	}
+	if strings.Contains(attachMode, "tty-in") {
```
Moved to comma-splitting and case-switching. Also, grouped bools in an anonymous struct.

lucab commented on stage0/app.go:
```diff
@@ -232,6 +232,28 @@ func AddApp(cfg AddConfig) error {
+			continue
+		}
+		pm.Annotations.Set(k, v)
+	}
```
See TODO at the top. But yes, I'll start tackling it here.

lucab commented:
PTAL. I've fixed the issues reported in previous reviews and cleaned up most of the the TODOs. There are still several in `iottymux.go` (especially in TTY code) but I don't want to block on those.

s-urbaniak commented on stage0/app.go:
```diff
@@ -30,6 +30,7 @@ import (
 	"github.com/coreos/rkt/common/apps"
 	pkgPod "github.com/coreos/rkt/pkg/pod"
 	"github.com/coreos/rkt/pkg/user"
+	//	stage1types "github.com/coreos/rkt/stage1/common/types"
```
superfluous

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+
+	// as soon as one output copying routine fails, this unblocks and the whole process exits
+	err = <-c
+	if err == io.EOF {
```
I'd leave this assertion to the caller's side.

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+	// TODO(lucab): send fd to systemd-manager to ensure we never lose the PTY master fd
+
+	// signal to systemd that the PTY is ready and application can start
+	ok, err := daemon.SdNotify(true, "READY=1")
```
the `err` is not checked, only `ok`

lucab replied:
This should be in line with SdNotify signature: https://github.com/coreos/go-systemd/blob/v14/daemon/sdnotify.go#L28-L31

In this case we require sd-notification to proceed, so message non-delivery is an hard failure here.

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+// It is never canceled explicitly, as it is bound to the lifetime of the main process.
+func acceptConn(socket net.Listener, c chan<- net.Conn, stream string) {
+	for {
+		conn, err := socket.Accept()
```
As far as I see we never close this socket. I guess when the process quits, the socket will be closed implicitly, but in case of an error implying process exit we could close it.

lucab replied:
The socket itself should be closed by https://github.com/coreos/rkt/pull/3396/files/3dddf85bfc65ad49812f8ac33e026c52bcb255bd#diff-9db502f4287233e5b508da56fd20e896R381 (but in the TTY codepath it is missing, right).

lucab replied:
fixed (at caller location).

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+			dst.Close()
+			ec <- io.EOF
+		}
+	}
```
I'd just make this two separate anonymous funcs

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+	rd := bufio.NewReader(src)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if len(lineOut) == 0 && err != nil {
```
what about the case if `len(lineOut) > 0 && err != nil` which can happen according to the doc?

s-urbaniak replied:
i.e. we could miss some err

lucab replied:
My understanding is that in this case we got a short read: we flush the incomplete line to the channel and proceed to the next read, until we get an hard failure (len=0 and error).

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+		case l := <-lines:
+			// TODO(lucab): ordered non-blocking writes
+			for i := len(logs) - 1; i >= 0; i-- {
+				out := fmt.Sprintf("%s %s %s", time.Now().Format(time.RFC3339Nano), streamLabel, l)
```
For hygiene we should dependency inject the formatting function having a `func(string) string` signature.

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,573 @@
+				_, err := conns[i].Write(l)
+				if err != nil {
+					conns[i].Close()
+					conns = conns[:i]
```
same comment as above

lucab commented:
Some more cleanup and rebased against some new conflicts.

squeed commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,598 @@
+	return err
+}
+
+// actionAttach handles TTY muxing and proxying.
```
actionTTYMux?

lucab replied:
Yep. Other docstrings were also wrong due reworks, fixed them as well.

squeed commented on stage1/init/common/units.go:
```diff
@@ -162,15 +162,7 @@ func ImmutableEnv(p *stage1commontypes.Pod, interactive bool) error {
-			opts = append(opts, unit.NewUnitOption("Service", "StandardOutput", "journal+console"))
-			opts = append(opts, unit.NewUnitOption("Service", "StandardError", "journal+console"))
-		}
+		opts := w.SetupAppIO(p, ra, binPath, interactive, p.Debug)
```
Can you move this in to AppUnit?

lucab replied:
Move, thanks for the refactoring. It also makes app-add support cleaner.

lucab commented:
Rebased after recent units/runc cleanup. PTAL.

s-urbaniak commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,599 @@
+		return err
+	}
+	err = json.NewEncoder(sf).Encode(endpoints)
+	_ = sf.Sync()
```
same here, I don't think that is needed, it won't hurt though.

lucab replied:
You are right, those are not strictly needed. Removing to reduce the noise.

lucab commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,599 @@
+	}
+
+	// as soon as one output copying routine fails, this unblocks and the whole process exits
+	return <-c
```
This codepath is reached from stage0 (`rkt attach`) and all errors here mean the app is gone so further re-attach will fail at connection time. I've left a note to split this logic a bit better, as it is no immediately clear what's the context around this snippet.

lucab commented on stage1/init/common/units.go:
```diff
@@ -195,6 +182,146 @@ func ImmutableEnv(p *stage1commontypes.Pod) error {
+//
+// For complete details, see dev-docs at Documentation/devel/log-attach-design.md
+func (uw *UnitWriter) SetupAppIO(p *stage1commontypes.Pod, ra *schema.RuntimeApp, binPath string, opts ...*unit.UnitOption) []*unit.UnitOption {
+	if p.Interactive {
```
Good catch!

lucab commented:
@s-urbaniak thanks for the comment, I added some changes to address them. I think we are close to the final round, @squeed mind going for it?

squeed commented on stage1/init/common/units.go:
```diff
@@ -195,6 +182,150 @@ func ImmutableEnv(p *stage1commontypes.Pod) error {
+		return opts
+	}
+
+	flavor, systemdVersion, err := GetFlavor(uw.p)
```
Should we add this to RuntimePod? Probably

squeed replied:
(Just bikeshedding here)

lucab replied:
I'm not sure we have that information when we flush RuntimePod to disk.

squeed commented on stage1/init/common/units.go:
```diff
@@ -195,6 +182,150 @@ func ImmutableEnv(p *stage1commontypes.Pod) error {
+		return opts
+	}
+	// Attach needs https://github.com/systemd/systemd/pull/4179, ie. systemd v232 or a backport
+	if ((flavor == "src" || flavor == "host") && systemdVersion < 232) ||
```
Does it make sense to error out if the annotations don't match the systemd version?

lucab replied:
We probably don't gain much from that, but I prefer explicit error as a general approach. Added it.

squeed commented on stage1/init/common/units.go:
```diff
@@ -595,6 +730,28 @@ func (uw *UnitWriter) AppReaperUnit(appName types.ACName, binPath string, opts .
 	)
 }
 
+// AppSocketUnits writes all socket units for the given app in the given path.
```
"Writes a socket unit"

lucab replied:
Those are typically more than one, likely 3 (in, out, err).

squeed replied:
But the method is called once per socket. The method name makes sense, just not the docblock.

lucab replied:
Ah, you are right, amended it.

squeed commented on stage1/iottymux/iottymux.go:
```diff
@@ -0,0 +1,601 @@
+
+// Targets references all attachable endpoints, for status persistence
+type Targets struct {
+	Version int        `json:"version"`
```
I'm sure this version doesn't hurt anything, but given that iottymux is embedded in the pod, it's probably not necessary.

lucab replied:
This JSON however is passed back to stage0 and to user  (in list mode). At some point it should unified with https://github.com/coreos/rkt/issues/3551.

squeed commented on tests/rkt_attach_test.go:
```diff
@@ -0,0 +1,102 @@
+
+// TestAttachSmoke is a smoke test for rkt-attach. It exercises several
+// features: tty/streams mux, auto-attach, interactive I/O.
+func TestAttachSmoke(t *testing.T) {
```
Hmm, can you also add a test that uses the sandbox?

lucab replied:
Definitely right, added.

squeed commented:
A question I had after reading the docs: does an application with `--stdout=stream` also log to the journal?

lucab commented:
No, at the moment `stream` and `tty` don't log to journal. But that is probably going to change as soon as we touch into logmode support for k8s.

lucab commented:
@squeed I added a bunch of changes for your comments and a test for sandbox-attach (killing some bugs while at it).

thomasfricke commented:
@lucab @s-urbaniak preparing a training for I tried to work with rkt in minikube.

minikube start --vm-driver=kvm --container-runtime=rkt --network-plugin=cni


The version

rkt Version: 1.14.0

is quite outdated. Do you have some newer version, as the old one has
the issue with attaching to the tty and the networking inside does not
work?

I would like to use rkt as an alternative in demos and training, so it
would be nice if the version coming with minikube would work at least.

This command shoud work out of the box
kubectl run  busybox --image=busybox sh

At the moment I work around this issue with 
kubectl run  busybox --image=busybox sleep 1000

and 

kubectl exec -it busybox-434806045-3qcgt sh

This is ugly for demos. Any idea or hint?

Thanks in advance!

jonboulle commented:
@thomasfricke have you filed an issue against minikube about this?

Thomas Fricke <notifications@github.com> schrieb am Mi., 25. Jan. 2017,
15:20:

> @lucab <https://github.com/lucab> @s-urbaniak
> <https://github.com/s-urbaniak> preparing a training for I tried to work
> with rkt in minikube.
>
> minikube start --vm-driver=kvm --container-runtime=rkt --network-plugin=cni
>
> The version
>
> rkt Version: 1.14.0
>
> is quite outdated. Do you have some newer version, as the old one has
> the issue with attaching to the tty and the networking inside does not
> work?
>
> I would like to use rkt as an alternative in demos and training, so it
> would be nice if the version coming with minikube would work at least.
>
> This command shoud work out of the box
> kubectl run busybox --image=busybox sh
>
> At the moment I work around this issue with
> kubectl run busybox --image=busybox sleep 1000
>
> and
>
> kubectl exec -it busybox-434806045-3qcgt sh
>
> This is ugly for demos. Any idea or hint?
>
> Thanks in advance!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/rkt/pull/3396#issuecomment-275119304>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACewNwlV9QPMYUwSHqeCoZWfQf0rBNL-ks5rV1oUgaJpZM4K4aqg>
> .
>

s-urbaniak commented:
@thomasfricke I created https://github.com/kubernetes/minikube/issues/1042

s-urbaniak commented:
resource cleanup is not properly handled upon `SIGTERM`, I'll submit a fix and a test.

s-urbaniak commented:
fix for the above is available under https://github.com/s-urbaniak/rkt/commit/cb346c636643d64e8b3b3e461d61399cc71dfeb6

s-urbaniak commented on Documentation/devel/log-attach-design.md:
```diff
@@ -0,0 +1,354 @@
+
+This component takes care of multiplexing TTY and receiving clients for attaching.
+
+It is started as an instance of the templatedd `ttymux@.service` service by a `After=` dependency from the application.
```
"of the templated"

lucab replied:
fixed.

lucab commented:
@s-urbaniak Thanks, I picked up your commit in this branch.

lucab commented:
ok to test

lucab commented:
Green, merging.
